cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,122 +1266,742 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+ * special handling has to happen in the calling code.
+ *
+ * @param input table_view of source table to be split
+ * @param splits the splits specified by the user, or an empty vector if no splits
+ * @returns true if the input is empty, false otherwise
+ */
+bool check_inputs(cudf::table_view const& input, std::vector<size_type> const& splits)
+{
+  if (input.num_columns() == 0) { return true; }
+  if (splits.size() > 0) {
+    CUDF_EXPECTS(splits.back() <= input.column(0).size(),
+                 "splits can't exceed size of input columns");
+  }
+  size_type begin = 0;
+  for (auto end : splits) {
+    CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
+    CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
+    CUDF_EXPECTS(end <= input.column(0).size(), "Slice range out of bounds.");
+    begin = end;
+  }
+  return input.column(0).size() == 0;
+}
+
+};  // anonymous namespace
+
+namespace detail {
+
+/**
+ * @brief A helper struct containing the state of contiguous_split, whether the caller
+ * is using the single-pass contiguous_split or chunked_pack.
+ *
+ * It exposes an iterator-like pattern where contiguous_split_state::has_next()
+ * returns true when there is work to be done, and false otherwise.
+ *
+ * contiguous_split_state::contiguous_split() performs a single-pass contiguous_split
+ * and is valid iff contiguous_split_state is instantiated with 0 for the user_buffer_size.
+ *
+ * contiguous_split_state::contiguous_split_chunk(device_span) is only valid when
+ * user_buffer_size > 0. It should be called as long as has_next() returns true. The
+ * device_span passed to contiguous_split_chunk must be allocated in stream `stream` by
+ * the user.
+ *
+ * None of the methods are thread safe.
+ */
+struct contiguous_split_state {
+  contiguous_split_state(cudf::table_view const& input,
+                         std::size_t user_buffer_size,
+                         rmm::cuda_stream_view stream,
+                         rmm::mr::device_memory_resource* mr,
+                         rmm::mr::device_memory_resource* temp_mr)
+    : contiguous_split_state(input, {}, user_buffer_size, stream, mr, temp_mr)
+  {
+  }
+
+  contiguous_split_state(cudf::table_view const& input,
+                         std::vector<size_type> const& splits,
+                         rmm::cuda_stream_view stream,
+                         rmm::mr::device_memory_resource* mr,
+                         rmm::mr::device_memory_resource* temp_mr)
+    : contiguous_split_state(input, splits, 0, stream, mr, temp_mr)
+  {
+  }
+
+  contiguous_split_state(cudf::table_view const& input,
+                         std::vector<size_type> const& splits,
+                         std::size_t user_buffer_size,
+                         rmm::cuda_stream_view stream,
+                         rmm::mr::device_memory_resource* mr,
+                         rmm::mr::device_memory_resource* temp_mr)
```

Comment:
Would it make sense to make this constructor private? That way the API of this struct is very clear that in can only be instantiated in the one of two modes. Right now you could construct it with both splits and `user_buffer_size`, which would result in erroneous behavior I assume (but have not verified).

Reply:
Yes it would. Warning, I am going to move some more of this to `private`, that was the intent and it looks like I missed doing this.

Reply:
except for `compute_batches`:
```
error: The enclosing parent function ("compute_batches") for an extended __device__ lambda cannot have private or protected access within its class
```